Python3 support

xiaket opened the pull request:
Hi everyone, 

This is at this stage an **EARLY EXPERIMENTAL** PR for python3 support. All unit tests passed locally both in Python 2.7.14 and 3.6.5.

I've utilized [future](http://python-future.org) do the stupid changes, and it's part of the stacker code base now. In future, if we want to drop python2 support, we can simply remove it from our code base.

To make it easier to run the tests in 2 environments, I've created two virtual environments and they are managed by a bash script embedded in this PR. Here's a little help for the script(`manage`) here:

To run a linting test for python2: `./manage lint2` (or `./manage l2` for short)
To run a nose test for python3: `./manage nose3` (or `./manage n3` for short)
To run everything: `./manage run`

It's a handy tool created to make the development easier. Technically, it should not be a part of this PR, but for now, I decide to let it in so you do not have to set up a test environment.

This is a huge PR, I've tried my best to make each commit atomic, and I hope I've explained the rationale for the change in the commit message. If you have any questions, just leave a message here or shoot me a message on Slack. Thanks.

xiaket commented:
All tests will fail, because I left Makefile unchanged, and essentially all code will import `future` and cause an error. I hope that's not an issue for now.

ejholmes commented:
I may not get a chance to review this for a while, but just wanted to say this is awesome and thanks for opening the PR @xiaket.

For what it's worth, I definitely think we should start getting python3 tested (maybe just running the full test suite against python3 on the master branch), so that we can get it green, and then keep it green.

/cc @phobologic

phobologic commented:
This is awesome - thanks so much @xiaket - I'll try to get a look soon!

phobologic commented:
Pinging @wmrsr as well, since he's poked at getting 3.x working before as well.

xiaket commented:
FYI, I've done another change and hopefully it will make all the tests pass in CircleCI.

One other thing I want to report is I can see a 100% speed boost when we run the tests in Python 3:

```
# running unit test in python2
Ran 343 tests in 22.320s
# running unit test in python3
Ran 343 tests in 11.238s
```

xiaket commented:
I configured CircleCI to run the unit-tests and functional-tests in Python3, all the tests are green now, I'll take a look at those skipped tests in bats(lack of AWS credentials) tomorrow. After that, I'll use [pyannotate](https://github.com/dropbox/pyannotate) to add type hinting to the codebase on a personal branch, so we could identify some more bugs lurking in the codebase and remove some of the type-checking logic I've added in this PR.

Please note that I've configured CircleCI to only run lint test against Python 3, I believe it is a sensible choice.

xiaket commented:
Fixed a few issues in bats tests. I haven't checked the last ones that uses the Role tho.

Added Python 3.5 to the picture and one of the tests failed, will look into it.

xiaket commented:
I've identified two bugs in the current code base by running unit tests under Python 3.5:

The first one is described in f6733af, essentially our current implementation cannot handle a corner case.

I haven't find a good way to deal with the second bug yet (which occasionally will cause the tests to fail [in Python 3.5](https://circleci.com/gh/xiaket/stacker/67?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link)): the `yaml_to_ordered_dict` in `stacker/utils.py` cannot guarantee the order specified in yaml. 

We have a test for this behaviour in `stacker/tests/test_util.py`, it will pass both in Python 2.7 and Python 3.6, however, if we replace all occurrences of `hook2` to `sss`, we should expect that the test would pass, but it will fail, even in Python 2.7. I think we need that ordering behavior, I'll try to come up with a solution later.

russellballestrini commented:
@xiaket that YAML logic is a tricky beast, I was the last to touch it, let me know if you want to talk through it or anything over Slack.

xiaket commented:
Thanks for offering your help @russellballestrini ! Yeah, that piece of logic is a little hard to understand,  it would be good to get some help on this one, can you please review the change I've done in 2ba8d50? Thanks!

russellballestrini commented:
@xiaket your changes in that commit look great and the level of detail in the commit message is excellent. Very cool!

phobologic commented on stacker/tests/actions/test_diff.py:
```diff
@@ -137,7 +140,8 @@ def test_normalize_json(self):
             '}\n'
         ]
-        self.assertEquals(normalized_template, normalize_json(template))
+        for i, line in enumerate(normalize_json(template)):
```
This looks like functionality change in the tests - was this required for python3?

xiaket replied:
Sorry for not describing it clearer in the commit message. I'll explain my rationale here.

The template that we see here is generated by `normalize_json`, where it uses `json.dumps` to dump a data structure. In Python 2, the default separator of the `dumps` call is `', '`, hence the extra trailing space. In Python 3, this behavior is fixed and the json does not have that extra trailing space.

On second thought, I think the best approach here is restore the old way to run tests (`self.assertEquals(normalized_template, normalize_json(template))`), but remove all the trailing space in the template in the test fixture, meanwhile, use `separators=(',', ': ')` in the `json.dumps` call to suppress the trailing space and unify the behavior in Python 2 and Python 3.

phobologic commented on stacker/tests/lookups/handlers/test_kms.py:
```diff
@@ -9,18 +13,20 @@
         with mock_kms():
             decrypted = handler(self.secret)
-            print "DECRYPTED: %s" % decrypted
+            print("DECRYPTED: %s" % decrypted)
```
This looks like a leftover debug statement - we can probably remove it, though not necessary here.

xiaket replied:
OK, I'll remove it.

xiaket commented on Makefile:
```diff
@@ -4,17 +4,23 @@ build:
 lint:
-	flake8 --exclude stacker/tests/ stacker
-	flake8 --ignore N802 stacker/tests # ignore setUp naming
+	flake8 --ignore E402 --exclude stacker/tests/ stacker
```
Because future monkey patched some of the builtins for compatibility reasons. The code would look like this(except from `stacker/config/__init__.py`):

```
from __future__ import print_function
from __future__ import division
from __future__ import absolute_import
from future import standard_library
standard_library.install_aliases()
from builtins import str
import copy
import sys
```

flake8 would consider the last three imports should happen before `standard_library.install_aliases()`, wherein this case, since we are monkey patching, we cannot do what flake8 suggests.

We could alternatively disable E402 check in the code by writing `# noqa: E402` after the `install_aliases` call. Currently, I'm choosing the lazy way of disabling that check globally. I can add all these `noqa` lines if you want. :)

phobologic replied:
Oh, sorry, I was confused - mixed it up with N802 above.  This makes sense, and I think I'm ok just keeping it like you have it now.  Thanks for the explanation.

xiaket commented on manage:
```diff
@@ -0,0 +1,87 @@
+#!/bin/bash
+
```
I think we need to remove this script before we merge it in. It won't stay for long.

xiaket commented on stacker/actions/base.py:
```diff
@@ -1,3 +1,7 @@
+from __future__ import print_function
```
Yeah, I think we do. Think about the case where we are debugging locally and we added a print statement to add some ad-hoc output, this statement will make us aware that we should be using the print function. These imports will ask us all to write Python 3 compatible code from the start.

phobologic replied:
Ok, that works - be curious how we enforce this going forward.  If someone adds a new module and forgets this, would tests somehow figure it out?

xiaket replied:
We can add a lint rule there. :D

xiaket commented on stacker/stack.py:
```diff
@@ -72,7 +76,7 @@ def __init__(self, definition, context, variables=None, mappings=None,
         self.enabled = enabled
         self.protected = protected
-        self.context = copy.deepcopy(context)
+        self.context = context
```
I really don't know what I can do here, without this change, tests in Python 3 will not pass, complaining something cannot be pickled. I've done some research, it might have something to do with our use of `schematics`.

Can we have a good test case and fix the memory issue instead?

phobologic replied:
#111 is the original issue, though strangely I think my fix (that is linked in that issue) was rolled back by #550.  @ejholmes do you remember why you might have rolled it back, and why it no longer causes issues with memory growth?  It might be the memory issue has been resolved entirely because of the DAG work.

phobologic replied:
Ok, I created a quick script to test if memory usage changed after removing the deepcopy:

```python
import sys
import resource

from stacker.context import Context
from stacker.config import Config, Stack


def generate_context(stacks, conf):
    stack_names = [d["name"] for d in stacks]
    stack_objs = [Stack(d) for d in stacks]

    conf.stacks = stack_objs
    conf.validate()
    return Context(
        environment={},
        stack_names=stack_names,
        config=conf,
    )


conf = Config()
conf.namespace = "test-namespace"

stacks = []
STACK_COUNT = int(sys.argv[1])

for i in range(STACK_COUNT):
    stacks.append({"name": "vpc%s" % i, "class_path": "blueprints.VPC"})


start_mem = resource.getrusage(resource.RUSAGE_SELF).ru_maxrss
for i in range(len(stacks)):
    ii = i + 1
    s = stacks[:ii]
    ctx = generate_context(s, conf)

stacks_created = ctx.get_stacks()

end_mem = resource.getrusage(resource.RUSAGE_SELF).ru_maxrss
total_mem = end_mem - start_mem

print total_mem,
```

And then ran it repeatedly, increasing the config size by one stack each time with this shell script, and dumped the output into a csv format file:

```shell
#!/bin/bash

ROWS=""

for i in $(seq 100)
do
    echo "$i,$(python ./test_stacker_mem.py $i)"
done
```

That resulted in this graph:

<img width="953" alt="screenshot 2018-06-23 15 36 16" src="https://user-images.githubusercontent.com/517334/41814200-3fccb500-76fb-11e8-994d-5d43cfeb6e1b.png">

So it looks like in 1.3 we regressed and started using more memory once more, and this should help fix it.  @ejholmes unless you can think of a reason this might be bad, I'm going to go ahead and go witht his too.

(BTW, here's the google spreadsheet with my run's results: https://docs.google.com/spreadsheets/d/15rP-RPYg7DtFQIwB_lK6oaJwPQHG6tz1NfxsvUKVJeM/edit?usp=sharing)

ejholmes replied:
We already talked about this offline, but I feel like this should be fine, and removing the deepcopy is the right thing to do. May have some unintended side effects for anyone doing something crazy `self.context` in custom blueprints.

xiaket commented:
@phobologic Thanks for your response! It does give me a chance to re-think what I had done, and it turns out, some of the decisions that I made earlier might be wrong.

By the way, I have a question to ask. Currently, I'm working on the type checking of the stacker codebase, so for example when a function receives both `str` and `unicode` for a parameter, we know that it's not good and we should fix the caller of the function. However, this will bring another round of automatic change to the codebase, moreover, some people might not like that type declaration in the function signature, because it is highly opinionated.

My current approach is to add the type checking in a personal branch, identify all the problematic function calls and fix them, backport all the change to python3 branch and push them here. But then we will lose some of the visibility.

Do you think it's a good approach? What's your opinion of doing type checking? Should we have another PR for this type-checking change (because this PR is already bulky)?

phobologic commented on stacker/actions/build.py:
```diff
@@ -108,7 +112,7 @@ def _resolve_parameters(parameters, blueprint):
     param_defs = blueprint.get_parameter_definitions()
 
-    for key, value in parameters.items():
+    for key, value in list(parameters.items()):
```
Awesome, thanks, I can do a once over after you get a chance to look at it to help find any that might get missed.

phobologic commented:
@xiaket I think it'd be best to start a branch from this branch for your type checking, and leave this PR just about getting python 3 compatibility working for the reason you specified - it's already a pretty big PR.  There might be a better way to handle this - @ejholmes pinging you since you might have opinions/ideas.

xiaket commented:
Hi, @phobologic , I've done my work on type checking and I didn't find any major issues that block us from getting this PR merged. I've also cleaned up some of the mess that futurize had done. Hope everything make sense now(especially the part where I added a flake8 plugin to force the future imports, 71857c4).

I went through the convo and as far as I see, we have these pending issues in this PR:

- removal of deepcopy of the context object in `stacker/stack.py`.
- re-implementation of dag reduction in f6733af and 9132809.

troyready commented:
@xiaket how would you feel about switching to https://pypi.org/project/formic2/ (https://github.com/wolfhong/formic) instead of the current conditional formic/formic-py3 install?

That 'formic2' fork seems to have gotten better/recent attention.

xiaket commented:
@troyready I think it's a good idea, I'd love to make the change and run the tests, anyone else on board?

phobologic commented:
@troyready @xiaket formic2 sounds good to me!

phobologic commented on stacker/config/__init__.py:
```diff
@@ -87,12 +93,18 @@ def render(raw_config, environment=None):
     try:
-        buff.write(t.substitute(environment))
-    except KeyError, e:
+        substituted = t.substitute(environment)
```
I think we can clean this up some:

```
try:
    substituted = t.substitute(environment)
except KeyError as e:
    raise exceptions.MissingEnvironment(e.args[0])
except ValueError:
    substituted = t.safe_substitute(environment)

if not isinstance(substituted, str):
    substituted = substituted.decode("utf-8")

buff.write(substituted)
```

xiaket replied:
I'll see what I can do.

xiaket replied:
Yeah, you are right, this will do.

However, when I was reading this part one more time, I realized that there could be another issue here. This try-except block might not always work as what we had expected it to do. substitute will always run sequentially, and it will panic at the first sight of error. So if we have a config file/environment values pair, that potentially will throw a `KeyError` and a `ValueError`, but the `KeyError` will only occur after the `ValueError`, then `KeyError` will not be raised.

Anyway, this is just a side note, and I think it does not fit into this PR in any way.

phobologic commented on stacker/dag/__init__.py:
```diff
@@ -168,22 +173,27 @@ def transitive_reduction(self):
-                    if xz and xy and yz:
-                        logger.debug("Removing edge %s -> %s" % (x, z))
-                        graph[x].remove(z)
+        combinations = []
```
ping @ejholmes  :)

xiaket replied:
@phobologic Hi Michael, thanks for you interest in this new DAG implementation. As I've stated in f6733af3, the previous implementation is flawed in that it cannot handle some of the corner cases.

Take a real world example here, software A depends on B and D, while B depends on C and C depends on D. What this `transitive_reduction` does is to simply the dependency tree to `A -> B -> C -> D`, while the previous implementation cannot return the correct result.

As for the implemetation, the idea is with all current edges(`A -> B`, `A -> D`, `B -> C`, `C -> D`), we will join the edges to create a new edge if the end of one edge is the beginning of another, so for `A -> B` and `B -> C`, we'll create `A -> C`, and, since we are running in a loop, we'll use `A -> C` and `C -> D` to create `A -> D` and all that. After this loop, we have marked all the connectivity between nodes. Then we'll mark all those newly constructed edges and remove the unnecessary nodes.

I've discovered this flaw during a Python 3.5 test run, where the key order of dictionaries are different from Python 2.7 and Python 3.6.

As for performance, correct me if I'm wrong but I think both implementation are O(N^3), and I don't believe this could be an issue since we are using DAG to handle Cloudformation templates, which will never be huge.

phobologic commented on stacker/tests/test_dag.py:
```diff
@@ -119,9 +122,9 @@ def test_downstream():
-    assert dag.all_downstreams('a') == ['c', 'b', 'd']
-    assert dag.all_downstreams('b') == ['d']
-    assert dag.all_downstreams('d') == []
+    assert set(dag.all_downstreams('a')) == set(['c', 'b', 'd'])
```
What's the reason for converting all of these to sets?  I don't think it changes the behavior really, but not sure why it's necessary.  Is this a python3'ism?

xiaket replied:
Yeah, you are right, this change was made when I was running the test code and I do have seen `dag.all_downstreams('a')` turned into `['b', 'c', 'd']`, which failed the test. I apologize because I haven't dug deep enough to see what's going on and used `set` to mute the failure.

I'll look into this.

xiaket replied:
Yeah, you are right, this change was made when I was running the test code and I do have seen `dag.all_downstreams('a')` turned into `['b', 'c', 'd']`, which failed the test. I apologize because I haven't dug deep enough to see what's going on and used `set` to mute the failure.

I'll look into this.

phobologic commented on tests/suite.bats:
```diff
@@ -100,35 +100,7 @@ EOF
-    }
-}
-EOF
+  assert $(echo "$output" | grep -v "Using default" | python -c "import sys, json; data = json.loads(sys.stdin.read()); print(data['steps']['vpc']['deps'] == [] and data['steps']['bastion1']['deps'] == ['vpc'] and data['steps']['app2']['deps'] == ['bastion2'])") = 'True'
```
What necessitated the changes in this file?  Are they related, or is this just a change?

danielkza commented:
This is very nice work, I have been looking to move all our tools to Python 3 ASAP, and this will help a lot :)

I noticed that there are some custom scripts to run tests on multiple Python versions, and I have always liked using [tox](https://tox.readthedocs.io/) for that purpose. It's battle-tested and likely familiar to any Python developers that want to start working in Stacker.

xiaket commented:
@troyready @phobologic Sadly use of formic2 will fail another unit test for all Python versions. I'll revert the change.

xiaket commented:
@danielkza Thanks for sharing that! I had bumped into tox in the past but I didn't know how to use it. This will help a lot in my work. 👍

danielkza commented:
@xiaket I have good experience with tox, if you need any support you can just ask on the stacker Slack channel and I'll do my best to help :)

troyready commented:
@xiaket it needs a small change to make formic2 work, sorry I didn't catch that before:
```
diff --git a/setup.py b/setup.py
index 7b04581..88dbd5a 100644
--- a/setup.py
+++ b/setup.py
@@ -11,7 +11,7 @@ install_requires = [
     "boto3>=1.3.1",
     "PyYAML~=3.12",
     "awacs>=0.6.0",
-    "formic~=0.9b",
+    "formic2~=1.0",
     "gitpython~=2.0",
     "schematics~=2.0.1",
 ]
diff --git a/stacker/hooks/aws_lambda.py b/stacker/hooks/aws_lambda.py
index 245d1bf..b96c4db 100644
--- a/stacker/hooks/aws_lambda.py
+++ b/stacker/hooks/aws_lambda.py
@@ -6,7 +6,6 @@ import hashlib
 from StringIO import StringIO
 from zipfile import ZipFile, ZIP_DEFLATED
 import botocore
-from functools import partial
 import formic
 from troposphere.awslambda import Code
 from stacker.session_cache import get_session
@@ -119,9 +118,7 @@ def _find_files(root, includes, excludes, follow_symlinks):
 
     root = os.path.abspath(root)
     file_set = formic.FileSet(directory=root, include=includes,
-                              exclude=excludes,
-                              walk=partial(
-                                  os.walk, followlinks=follow_symlinks))
+                              exclude=excludes, symlinks=follow_symlinks)
 
     for filename in file_set.qualified_files(absolute=False):
         yield filename
```

xiaket commented:
Hey @ejholmes, this PR has been stuck here for quite a while, can you please squeeze some time to review some of the DAG related changes? Thanks!

phobologic commented:
Hey @xiaket - I went ahead and created a branch in the cloudtools/stacker repo to kick off functional tests, and it looks like we're running into issues with 2.7:

https://circleci.com/gh/cloudtools/stacker/1992

(here's the fake PR I made to kick off the functional tests: https://github.com/cloudtools/stacker/pull/610)

Any ideas?

phobologic commented:
Ok, after some creative debugging with print statements (and dealing with the OS X systems integrity protection, ugh), I think I've tracked it down to urllib in python2.7 having issues with byte strings:

```
   KeyError: 49
   [2018-06-24T11:20:55] vpc: failed (49)
   [2018-06-24T11:20:55] The following steps failed: vpc
### The following are debug print statements
DEBUG: urllib.quote: s arg: b'10.128.0.0/16'
DEBUG: urllib.quote: QUOTER var: <built-in method __getitem__ of dict object at 0x1099f3168>
   
DEBUG: botocore.vendored.requests.models._encode_params: data arg: {'Parameters.member.2.ParameterKey': 'PublicSubnets', 'Parameters.member.5.ParameterValue': 'NAT', 'StackName': u'stacker-test-namespace-vpc', 'Parameters.member.6.ParameterValue': 'false', 'Version': u'2010-05-15', 'Capabilities.member.1': 'CAPABILITY_NAMED_IAM', 'Parameters.member.5.ParameterKey': 'ImageName', 'Parameters.member.4.ParameterKey': 'BaseDomain', 'Parameters.member.7.ParameterValue': '10.128.0.0/16', 'Parameters.member.2.ParameterValue': '10.128.0.0/24,10.128.1.0/24,10.128.2.0/24,10.128.3.0/24', 'Parameters.member.1.ParameterValue': '', 'TemplateURL': u'https://s3.amazonaws.com/stacker-stacker-test-namespace/stack_templates/stacker-test-namespace-vpc/vpc-829da93c.json', 'Parameters.member.7.ParameterKey': 'CidrBlock', 'Tags.member.1.Value': u'stacker-test-namespace', 'Parameters.member.1.ParameterKey': 'InternalDomain', 'Parameters.member.3.ParameterValue': '10.128.8.0/22,10.128.12.0/22,10.128.16.0/22,10.128.20.0/22', 'Action': u'UpdateStack', 'Parameters.member.3.ParameterKey': 'PrivateSubnets', 'Tags.member.1.Key': 'stacker_namespace', 'Parameters.member.6.ParameterKey': 'UseNatGateway', 'Parameters.member.4.ParameterValue': '', 'Parameters.member.8.ParameterKey': 'InstanceType', 'Parameters.member.8.ParameterValue': 'm3.medium'}
   

DEBUG: botocore.vendored.requests.models._encode_params: RESULT var: [('Parameters.member.2.ParameterKey', 'PublicSubnets'), ('Parameters.member.5.ParameterValue', b'NAT'), ('StackName', 'stacker-test-namespace-vpc'), ('Parameters.member.6.ParameterValue', b'false'), ('Version', '2010-05-15'), ('Capabilities.member.1', 'CAPABILITY_NAMED_IAM'), ('Parameters.member.5.ParameterKey', 'ImageName'), ('Parameters.member.4.ParameterKey', 'BaseDomain'), ('Parameters.member.7.ParameterValue', b'10.128.0.0/16'), ('Parameters.member.2.ParameterValue', b'10.128.0.0/24,10.128.1.0/24,10.128.2.0/24,10.128.3.0/24'), ('Parameters.member.1.ParameterValue', b''), ('TemplateURL', 'https://s3.amazonaws.com/stacker-stacker-test-namespace/stack_templates/stacker-test-namespace-vpc/vpc-829da93c.json'), ('Parameters.member.7.ParameterKey', 'CidrBlock'), ('Tags.member.1.Value', 'stacker-test-namespace'), ('Parameters.member.1.ParameterKey', 'InternalDomain'), ('Parameters.member.3.ParameterValue', b'10.128.8.0/22,10.128.12.0/22,10.128.16.0/22,10.128.20.0/22'), ('Action', 'UpdateStack'), ('Parameters.member.3.ParameterKey', 'PrivateSubnets'), ('Tags.member.1.Key', 'stacker_namespace'), ('Parameters.member.6.ParameterKey', 'UseNatGateway'), ('Parameters.member.4.ParameterValue', b''), ('Parameters.member.8.ParameterKey', 'InstanceType'), ('Parameters.member.8.ParameterValue', b'm3.medium')]
```

It's specifically `urllib.quote` having an issue with the byte-string `b"10.128.0.0/16"` - it may have trouble with others as well, this is just the first byte string it's dealing with.

phobologic commented:
Hmm, nevermind, i was wrong about the address being the first byte string - there are others before it.

phobologic commented:
Ok, running into a little bit of a roadblock of lack of knowledge around strings/bytestrings in the python2/3 world.  I have to work on some other stuff, but I'll be around most of the day to chat (here or on Slack) if anyone gets a chance to take a look!

xiaket commented:
@phobologic Sorry for the late reply, I took the last two days fighting this (properly setup an environment alone took me one day) and I already have a dirty hack that can fix this issue, I'm looking for a clean solution ATM. I'll keep you updated about my progress.

phobologic commented:
Does the removal of importing the str method cause any other issues?  That doesn't look super dirty to me, though maybe I don't understand the implications for python3.  Thanks @xiaket!

xiaket commented:
@phobologic I believe the failure in dc10db4 is unrelated(throttling of API). I updated README and forced another round of tests, it's all green now.

phobologic commented:
@xiaket unfortunately the functional-tests in non-`cloudtools/stacker` based branches don't actually run (for security reasons).  The output shows each of the tests as `skipped` in circleci.

The good news is that it looks like after forcing retries (to get around the Throttling) it looks like the functional tests passed.  I'm going to merge this now!  Thanks!
